### PR TITLE
[refactor] #3833, #2373, #3601: Split off Kagami

### LINKED ISSUE #3833: [suggestion] Separate `validator` and `swarm` from Kagami
## Description

Kagami was originally designed as an accompanying tool for the Iroha codebase. It was used to generate configuration, cryptographic keys, and more. Mentioned extensively in Iroha 2 documentation, it instructs users on deploying Iroha 2. However, some added functionalities have now limited its usability in some scenarios.

### Commands overview

Kagami currently supports the following commands:

- `crypto`: Generate cryptographic key pairs using the given algorithm and either a private key or seed
- `schema`: Generate the schema for code generation in Iroha SDKs
- `genesis`: Generate the genesis block that is used in tests
- `config`: Generate the default client/peer configuration
- `docs`: Generate a Markdown reference of configuration parameters
- **`validator`**: Generate the default validator
- **`swarm`**: Generate Docker Compose configuration

While most commands are self-contained, `validator` and `swarm` are exceptions with specific dependencies.

### Issues

- `kagami validator`  only functions when CWD is the Iroha repo, leading to problems (e.g., #3741, #3700).
- `kagami swarm` shares the same limitations as it reuses `kagami validator`.
- Because of the complexity which `kagami validator` brought, [we discussed](https://github.com/hyperledger/iroha/issues/3603#issuecomment-1685919947) that it is probably good to keep Kagami as a fully internal utility. However, we benefit from the ability to mention it in public documentation.
### Proposed solution

To address these challenges, I propose to remove `validator` and `swarm` subcommands from Kagami. Thus, Kagami will remain a lightweight tool for both internal and external usage with ability to mention it in the documentation.

As for the `validator` and `swarm`, they could likely become separate binaries. The naming, location, and CLI for these new binaries is up to discussion.

### Related

- #2373

### PR DESCRIPTION
**Pull request by 0x009922**
## Description

Outline:

- **Refactor Kagami**
  - Remove `kagami validator`
  - Remove `kagami swarm`
  - Refactor `kagami genesis` according to the [comment](https://github.com/hyperledger/iroha/issues/3833#issuecomment-1693023658)
- **Introduce `iroha_swarm`**: according to the [comment](https://github.com/hyperledger/iroha/issues/2373#issuecomment-1690971061)
- **Introduce `iroha_wasm_builder_cli`**: a CLI around `iroha_wasm_builder`. It replaces `kagami validator` in the following way:
  ```bash
  # before
  kagami validator > validator.wasm
  
  # after
  iroha_wasm_builder_cli check ./default_validator
  iroha_wasm_builder_cli build ./default_validator --optimize --format > validator.wasm
  ```
- **Update CI scripts** which used the previous commands

### Linked issue

Closes #3833 #2373 #3601

### Benefits

- Kagami is now a portable and lightweight tool. There is no issue with mentioning it in the documentation.
- Swarm is now simpler, without redundant functionality. Easy to enhance in the future.
- A general tool for building smartcontracts. Easy to enhance in the future. More reliable than `kagami validator` thanks to spinners.

### Checklist

- [x] Investigate the logic behind the `format` option in `iroha_wasm_builder::Builder`, maybe open an issue to refactor it.
- [x] Self-review
- [x] Add README to `iroha_swarm`

### CODE REVIEW

File: wasm_builder_cli/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    Check {
+        #[command(flatten)]
+        common: CommonArgs,
+    },
+    /// Build the smartcontract
+    Build {
+        #[command(flatten)]
+        common: CommonArgs,
+        /// Enable smartcontract formatting using `cargo fmt`.
+        // TODO: why it is a part of `build` in wasm_builder?
```
Comment (0x009922) [inline wasm_builder_cli/src/main.rs:23]: Anyone who is in the context: could you provide a reasoning about this decision?

File: wasm_builder_cli/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+```bash
+iroha_wasm_builder_cli build path/to/project
+```
+
+**Build with options:**
+
+```bash
+iroha_wasm_builder_cli build path/to/project --optimize --format
+```
```
Comment (DCNick3) [inline wasm_builder_cli/README.md:23]: Maybe document how the `default_validator` should be built? It's a quite a common thing to want (for a developer).. AFAIU it's `iroha_wasm_builder_cli build default_validator --optimize > ./configs/peer/validator.wasm`

File: wasm_builder_cli/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                }
+            } else {
+                output
+            };
+
+            let bytes = output
+                .into_bytes()
+                .wrap_err("Failed to fetch the bytes of the output smartcontract")?;
+
+            let mut writer = BufWriter::new(stdout());
```
[the remark this answers is not in the document]
Comment (0x009922) [reply wasm_builder_cli/src/main.rs:100]: I tried to preserve the behaviour of Kagami, which outputs everything to stdout. Having an "output path" is not a problem for sure, and is better for the user, but it also introduces extra complexity by handling e.g. existence of the file. Swarm makes a prompt in this case.

File: wasm_builder_cli/README.md (hunk 2 above)
Comment (0x009922) [reply wasm_builder_cli/README.md:23]: Documented in `default_validator/README.md`. I am not sure this is a good location for it, but I haven't found any current documentation to patch.

File: tools/swarm/src/ui.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+}
+
+impl UserInterface {
+    pub(super) fn new() -> Self {
+        Self
+    }
+
+    #[allow(clippy::unused_self)]
+    pub(super) fn prompt_remove_target_file(&self, file: &AbsolutePath) -> Result<PromptAnswer> {
```
[the remark this answers is not in the document]
Comment (0x009922) [reply tools/swarm/src/ui.rs]: Refactored.

**Review by DCNick3 — DISMISSED** [review]

File: tools/kagami/src/genesis.rs [hunk 5]
```diff
@@ -58,46 +56,71 @@ pub enum Mode {
...
+impl TryFrom<ParsedValidatorArgs> for ValidatorMode {
+    type Error = color_eyre::Report;
+
+    fn try_from(value: ParsedValidatorArgs) -> Result<Self, Self::Error> {
+        let mode = match value {
+            ParsedValidatorArgs::Path(path) => ValidatorMode::Path(ValidatorPath(path)),
+            ParsedValidatorArgs::Inline(path) => {
+                let validator = ValidatorMode::Path(ValidatorPath(path))
+                    .try_into()
+                    .wrap_err("Failed to read the validator")?;
```
Comment (ilchu) [inline tools/kagami/src/genesis.rs:107]:
```suggestion
                    .wrap_err(format!("Failed to read the validator located at {}", path))?;
```

**Review by ilchu — COMMENTED** [review]
Good to feel the old crutches gone. Overall structure is much more digestible too.

File: wasm_builder_cli/src/main.rs (hunk 1 above)
Comment (DCNick3) [reply wasm_builder_cli/src/main.rs:21]: Why have it an option as `build` and not a separate action, like `cargo build` and `cargo fmt`?

Comment (ilchu) [reply wasm_builder_cli/src/main.rs:21]: Well it is a separate action that is run conditionally depending on whether this flag in the builder is set, isn't it?

Comment (0x009922) [reply wasm_builder_cli/src/main.rs:21]: I'm puzzled about the inclusion of `cargo fmt` in the build process. While it makes sense for `wasm_builder` to handle `cargo check` and `cargo build` due to specific rustc parameters, `cargo fmt` doesn't seem to need such configuration. So why is it part of the build? Is formatting not the developer's responsibility? Are there historical reasons for this design choice?

File: wasm_builder_cli/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                eyre!(
+                    "Failed to write the resulting file into {}",
+                    outfile.display()
+                )
+            })?;
+
+            eprintln!(
+                "✓ File is written into {}",
+                outfile.display().green().bold()
+            );
```
[the remark this answers is not in the document]
Comment (0x009922) [reply wasm_builder_cli/src/main.rs:108]: It is. I can certainly change this, but what do you think about usage of `spinoff::Streams::Stderr` for spinners?

File: tools/kagami/src/genesis.rs [hunk 7]
```diff
@@ -58,46 +56,71 @@ pub enum Mode {
...
+            inline_validator_from_file,
+            validator_path_in_genesis,
+            mode,
+        } = self;
+
+        let validator: ValidatorMode = match (inline_validator_from_file, validator_path_in_genesis) {
+            (Some(path), None) => {
+                eprintln!("WARN: You're using genesis with inlined validator.");
+                eprintln!(
+                    "Consider providing a validator in separate file with `--validator-path-in-genesis <PATH>`."
```
[the remark this answers is not in the document]
Comment (0x009922) [reply tools/kagami/src/genesis.rs]: Included the extended explanation into the `--help` message.

**Review by DCNick3 — APPROVED** [review]

File: Cargo.toml [hunk 8]
```diff
@@ -165,11 +165,13 @@ members = [
   "tools/kagami",
   "tools/kura_inspector",
   "tools/parity_scale_decoder",
+  "tools/swarm",
   "version",
   "version/derive",
   "wasm_codec",
   "wasm_codec/derive",
   "wasm_builder",
+  "wasm_builder_cli",
```
Comment (pesterev) [inline Cargo.toml:174]: Can we consider to move `wasm_builder_cli` to tools directory? My point is that `wasm_builder_cli` is a CLI tool.

Comment (0x009922) [reply Cargo.toml:174]: Sounds reasonable. Will do.

File: wasm_builder_cli/src/main.rs (hunk 1 above)
Comment (DCNick3) [reply wasm_builder_cli/src/main.rs]: Yeah, tbh, it does feel kind of unnecessary, especially if `cargo fmt` does not need any special flags

Comment (ilchu) [reply wasm_builder_cli/src/main.rs]: I'd suppose it is there to just make an all-inclusive package then. And maybe @Arjentix could clarify it more, but that's the gut feeling I have.

File: wasm_builder_cli/src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ilchu) [reply wasm_builder_cli/src/main.rs]: On a second look, std docs do state that stderr is for diagnostic usage as well, which is exactly what a spinner is. And one example I can come up with is that if user tries to redirect output, he wouldn't be happy seeing spinners contaminating useful output there.

File: wasm_builder_cli/src/main.rs (hunk 1 above)
Comment (0x009922) [reply wasm_builder_cli/src/main.rs]: In order to avoid piling up this PR, moved the discussion here: https://github.com/hyperledger/iroha/issues/3858

**Review by ilchu — APPROVED** [review]

**Review by DCNick3 — APPROVED** [review]

**Pull request merged**